### PR TITLE
[WIP] Implementation proposal

### PR DESCRIPTION
**Pull request by j1elo**
This is a proposal implementation of sdp-bridge. It is a work-in-progress, not thoroughly tested. The focus is on functionality, leaving architecture considerations aside. Specifically:

* Invalid scenarios are just caught and the process ends with a "BUG" message.
* Code is currently written for the WebRtcTransport, not for generic Transports.
* A static RtpCapabilities object (from `BrowserRtpCapabilities`) is used to represent the RTP caps of remote browsers. This should be worked on in the future, to obtain these caps from the SDP Answer that a browser send in response to an SDP Offer. Read more below.
* Some verbose logs are just commented out, instead of using more adequate technique of filtering by logging level.
* "mediasoup" and "mediasoup-client" are added as dependencies in package.json.

An important limitation we've found while writing this code is that there is a "logic loop" in the mediasoup API, which makes it impossible to create a proper SDP endpoint. This is a summary of the issue:

1. In order to send an SDP Offer, the offerer needs to include its own SSRC for the medias that are going to be sent.
2. In mediasoup, the SSRC is obtained from the `Consumer.rtpParameters.encodings[].ssrc`, so i.e. a Consumer must have been created beforehand.
3. In order to create a Consumer, you need to pass its constructor ([Transport.consume()](https://mediasoup.org/documentation/v3/mediasoup/api/#transport-consume)) an `RtpCapabilities` that describes the capabilities of the remote peer.
4. The `RtpCapabilities` that describe what a browser is capable of, would be obtained from an SDP Answer sent by it.
5. To make a browser send an SDP Answer, you have to first send it an SDP Offer. GOTO 1.

We're breaking this "loop" by hardcoding in the source code an RtpCapabilities that represents a typical SDP Answer from the browser, with 1 audio and 1 video stream. There are capabilities for Firefox, Safari, and Chrome, of which currently the Chrome ones are in use because they seem to be compatible with all other browsers, Firefox included.

Apart from that, another addition to the API was the `addConsumer` method. Just like sdp-bridge is able to generate an array `Producer[]` from an SDP Offer, it should also be able to generate an SDP Offer from an array `Consumer[]`, so it was necessary to add a method that allows providing these Consumers.

All in all, this only intends to be a preliminary implementation that opens up further design decisions and conversation.

### CODE REVIEW

**Comment by j1elo** [comment]
Please ignore notifications of WIP commits while we play with this

**Review by piranna — CHANGES_REQUESTED** [review]
This change is not needed at all, by using a npm `prepare` script [it will automatically build](https://docs.npmjs.com/cli/v6/using-npm/scripts#prepare-and-prepublish) the `lib/` folder version when using as a git dependency when doing `npm install`.

File: package.json [hunk 1]
```diff
@@ -15,12 +15,17 @@
     "type": "opencollective",
     "url": "https://opencollective.com/mediasoup"
   },
-  "main": "lib/index.js",
-  "types": "lib/index.d.ts",
+  "main": "lib/main.js",
+  "types": "lib/main.d.ts",
```
Comment (piranna) [inline package.json]: Standard filename is `index.js`, revert this.

File: src/main.ts [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+  // * processOffer
+  // * createAnswer
+
+  public async processOffer(sdpOffer: string): Promise<Producer[]> {
+    if (this.remoteSdp) {
+      console.error(
+        "[SdpEndpoint.processOffer] ERROR: A remote description was already set"
+      );
+      return [];
+    }
```
Comment (piranna) [inline src/main.ts]: RTCPeerConnection state machine allows to process an offer after having a remote SDP depending on what state we are, probably we should implement that too.

**Review by piranna — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (j1elo) [reply package.json]: Done

**Comment by j1elo** [comment]
Closing this until the fork is in a good shape for contribution:

https://github.com/OpenVidu/mediasoup-sdp-bridge

**Pull request closed without merging**